### ISSUE TITLE
Performance question

### ISSUE THREAD

**Issue by bgnorlov** [rethinkdb/rethinkdb#5593]
Can anyone explain why one query inside database slower than two queries in runtime? I have table 'folder' , table 'boards' with 'folders_id' index and table 'posts' with 'boards_id' index. My first query 
```javascript
r.connect(config.rethinkdb).then(dbConn=>{
  var time = Date.now();
  r.table('boards').getAll('1233131', {index: 'folders_id'}).concatMap(board=>{
    return r.table('posts').getAll(board('id'),{index: 'boards_id'})
  }).coerceTo('array').run(dbConn).then((res)=>{
    console.log(Date.now()-time, res.length);
  });
});
```
uses concatMap and it took nearly 151 ms. But secon query 
```javascript

r.connect(config.rethinkdb).then(dbConn=>{
  var time = Date.now();
  r.table('boards').getAll('1233131', {index: 'folders_id'}).map(r.row('id')).coerceTo('array').run(dbConn).then(ids=>{
    console.log(Date.now()-time, ids.length);
    var arr = ids;
    arr.push({index: 'boards_id'});
    var thisObj = r.table('posts');
    return thisObj.getAll.apply(thisObj, arr).coerceTo('array').run(dbConn)
  }).then((res)=>{
    console.log(Date.now()-time, res.length);
  });
});
```
uses two runtime queries (first retrieves all ids and second fetches all posts) and it took nearly 60 ms. I tested with other data and concatMap approach twice (or three times on small data) slower then approach with retrieves ids and then items itself

**Comment by chrisvariety** [comment]
Just a guess and the folks at RethinkDB would know more, but...

The first example runs a bunch of separate getAll's, one for each set of posts within each board, and the second just runs two getAll's total.

I'm not sure if you want to return _both_ boards _and_ posts or only _posts_. Assuming you want only posts, you can just do this which is roughly equivalent to your second example:

```
r.table('posts').getAll(r.args(
    r.table('boards').getAll('1233131', {index: 'folders_id'})('id').coerceTo('array')
  ), {index: 'boards_id'});
})
```

This will error out if there are zero posts with that board id FYI (see https://github.com/rethinkdb/rethinkdb/issues/2588 for a workaround)

**Comment by bgnorlov** [comment]
@chrisvariety Thanks for "r.args" trick - it gave 30% performance boost compare to approach of two separate request. By the way, I don't see a big difference - in both cases (nested concatMap or nested getAll in args) database need retrieve all boards in folder and all post for each board.

**Comment by danielmewes** [comment]
I think @chrisvariety 's explanation is the right one. Doing a single `getAll` with multiple keys is much more efficient than doing multiple `getAll`, one for each key. This is also at the core of our recent optimization to `eqJoin` which we shipped with RethinkDB 2.3.0. Previously `eqJoin` would do individual `getAll`s similar to your first query, while now it's combining multiple join keys into a single `getAll` request internally.
